packages/core/__tests__/annotations.test.tsx

```diff
@@ -0,0 +1,185 @@
+  })
+
+  it('popup is not rendered if annotation props are not passed', () => {
+    const exam = render(<Exam {...getExamProps()} />)
+    annotateText(exam, defaultTextToAnnotate)
+    expect(exam.queryByTestId('annotation-popup')).not.toBeInTheDocument()
+  })
+
+  it('callback is called when annotation is saved', () => {
+    const saveAnnotationMock = jest.fn()
+    const exam = render(
+      <Exam {...getExamProps()} annotations={{}} onClickAnnotation={() => {}} onSaveAnnotation={saveAnnotationMock} />
+    )
+    annotateText(exam, defaultTextToAnnotate)
+
+    const textArea = exam.getByTestId('edit-comment')
+    fireEvent.change(textArea, { target: { value: 'New Value' } })
+    fireEvent.click(exam.getByText('Vastaa'))
+
+    expect(saveAnnotationMock).toHaveBeenCalledTimes(1)
+    expect(saveAnnotationMock).toHaveBeenCalledWith({
+      annotationAnchor: defaultAnnotationAnchor,
+      annotationId: undefined,
+      displayNumber: '2',
+      length: defaultTextToAnnotate.length,
+      message: 'New Value',
+      selectedText: defaultTextToAnnotate,
+      startIndex: 67
+    })
+  })
+
+  it('annotations are added to dom when provided', () => {
+    function createAnnotation(id: number, startIndex: number, length: number): ExamAnnotation {
+      return {
+        annotationId: id,
+        annotationAnchor: defaultAnnotationAnchor,
+        hidden: false,
+        startIndex,
+        length,
+        message: '',
+        displayNumber: '',
+        selectedText: ''
+      }
+    }
+    const annotationProps: AnnotationProps = {
+      annotations: {
+        [defaultAnnotationAnchor]: [createAnnotation(1, 1, 1), createAnnotation(3, 4, 2)]
+      },
+      onClickAnnotation: () => {},
+      onSaveAnnotation: () => {}
+    }
+    const exam = render(<Exam {...getExamProps()} {...annotationProps} />)
+    expect(exam.container.querySelector('[data-annotation-id="1"]')).toBeInTheDocument()
+    expect(exam.container.querySelector('[data-annotation-id="3"]')).toBeInTheDocument()
+  })
+
+  it('grading instructions can be annotated', () => {
+    const gi = render(
+      <GradingInstructions
+        {...getCommonProps()}
+        annotations={{}}
+        onClickAnnotation={() => {}}
+        onSaveAnnotation={() => {}}
+      />
+    )
+    annotateText(gi, defaultTextToAnnotate)
+    expect(gi.getByTestId('annotation-popup')).toBeVisible()
+  })
+
+  it('attachments can be annotated', () => {
+    const gi = render(
+      <Attachments {...getCommonProps()} annotations={{}} onClickAnnotation={() => {}} onSaveAnnotation={() => {}} />
+    )
+    annotateText(gi, 'Mollit sint magna aliquip in.')
+    expect(gi.getByTestId('annotation-popup')).toBeVisible()
+  })
+
+  function annotateText(exam: RenderResult, text: string) {
+    mockWindowSelection(text, text.length)
+    const textElement = exam.getByText(text)
+    fireEvent.mouseDown(textElement)
+    fireEvent.mouseUp(textElement)
```

Comment:
voisi miettiä haluaisiko käyttää tätä `fireEvent`in sijaan https://testing-library.com/docs/user-event/intro

Reply:
aaa.. toi olis parempi. Katsotaan saanko käytettyä tuota